File: engines/dart/lib/src/rules/rules_engine.dart
import '../engine/game_system.dart';
import '../models/event.dart';
import '../models/game_definition.dart';
import '../models/game_state.dart';
import '../models/rule.dart';
import 'condition_evaluator.dart';
import 'effect_executor.dart';

/// Runs the cascade resolution loop (phase 5).
class RulesEngine {
  final List<RuleDef> _gameRules;
  final List<RuleDef> _levelRules;
  final ConditionEvaluator _condEval = ConditionEvaluator();

  RulesEngine(this._gameRules, this._levelRules);

  /// Evaluate all rules against the given events, cascading up to maxDepth.
  /// Also runs cascade-phase systems after each pass.
  List<GameEvent> evaluate(
    List<GameEvent> initialEvents,
    LevelState state,
    GameDefinition game,
    int maxDepth,
    List<GameSystem> cascadeSystems,
  ) {
    final allNewEvents = <GameEvent>[];
    var pendingEvents = List<GameEvent>.from(initialEvents);
    final effectExec = EffectExecutor(game);

    // All rules in priority order (game-level first, then level-local)
    final allRules = [..._gameRules, ..._levelRules];
    allRules.sort((a, b) => b.priority.compareTo(a.priority));

    for (int pass = 0; pass < maxDepth; pass++) {
      if (pendingEvents.isEmpty) break;

      final newEvents = <GameEvent>[];

      for (final event in pendingEvents) {
        for (final rule in allRules) {
          if (rule.on != event.type) continue;

          // Check once-fired
          if (rule.once && state.onceFiredRules.contains(rule.id)) continue;

          // Evaluate where condition
          if (!_condEval.evaluate(rule.where, event, state, game)) continue;

          // Evaluate if condition
          if (!_condEval.evaluate(rule.ifCond, event, state, game)) continue;

          // Fire rule
          if (rule.once) state.onceFiredRules.add(rule.id);

          for (final effect in rule.then) {
            final effectEvents = effectExec.execute(effect, event, state);
            newEvents.addAll(effectEvents);
          }
        }
      }

      // Run cascade-phase systems
      for (final sys in cascadeSystems) {
        final sysEvents =
            sys.executeCascadeResolution(pendingEvents, state, game);
        newEvents.addAll(sysEvents);
      }

      allNewEvents.addAll(newEvents);
      pendingEvents = newEvents;
    }

    return allNewEvents;
  }
}

File: engines/python/_rules_engine.py
"""
Rules engine (cascade resolution, phase 5).

Mirrors Dart's rules_engine.dart:
  - Evaluates all rules against pending events
  - Cascade systems run after each pass
  - Repeats up to maxCascadeDepth times
"""
from __future__ import annotations
from ._models import GameState
from ._game_def import GameDef
from . import _conditions as cond
from . import _effects as eff


class RulesEngine:
    def __init__(self, game_rules: list[dict], level_rules: list[dict]):
        self._all_rules = sorted(
            game_rules + level_rules,
            key=lambda r: r.get("priority", 0),
            reverse=True,  # higher priority first
        )

    def evaluate(
        self,
        initial_events: list[dict],
        state: GameState,
        game: GameDef,
        max_depth: int,
        cascade_systems: list,
    ) -> list[dict]:
        """
        Run cascade loop: for each pass, fire matching rules then cascade systems.
        Returns all newly emitted events (not including initial_events).
        """
        all_new_events: list[dict] = []
        pending = list(initial_events)

        for _ in range(max_depth):
            if not pending:
                break

            new_events: list[dict] = []

            for event in pending:
                for rule in self._all_rules:
                    if rule["on"] != event["type"]:
                        continue
                    # once-fired guard
                    if rule.get("once") and rule["id"] in state.once_fired_rules:
                        continue
                    # where condition
                    if not cond.evaluate(rule.get("where"), event, state, game):
                        continue
                    # if condition
                    if not cond.evaluate(rule.get("if"), event, state, game):
                        continue
                    # fire
                    if rule.get("once"):
                        state.once_fired_rules.add(rule["id"])
                    for effect in rule.get("then", []):
                        new_events.extend(eff.execute(effect, event, state, game))

            # cascade systems
            for sys in cascade_systems:
                new_events.extend(sys.execute_cascade_resolution(pending, state, game))

            all_new_events.extend(new_events)
            pending = new_events

        return all_new_events
